Generate macOS configuration profiles

joshspicer opened the pull request:
ref: https://github.com/microsoft/vscode/issues/148942

Generates a sample `.mobileconfig` profile, similar to how the ADMX file is generated for Windows. 

Also generates localized files following [this specification](https://github.com/ProfileManifests/ProfileManifests/wiki/Manifest-Format) to aid in producing customized `.mobileconfig` profiles.

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

joshspicer commented:
❗ Requires picking up https://github.com/microsoft/vscode-policy-watcher/pull/44

joshspicer commented on src/vs/code/electron-main/main.ts:
```diff
@@ -198,9 +198,16 @@ class CodeMain {
+			policyService = new NullPolicyService();
+		}
 		services.set(IPolicyService, policyService);
 
```
❗ Requires picking up https://github.com/microsoft/vscode-policy-watcher/pull/44

joshspicer replied:
https://github.com/microsoft/vscode/pull/241773/commits/96349ad28a5e21993d4f3ab708012d2b17b12c91

joshspicer committed:
generate configuration policy and manifests

joshspicer committed:
add darwin pipeline job

joshspicer committed:
argument for per-platform generation of policy

joshspicer committed:
adopt argument in pipeline

joshspicer committed:
formatting tweaks in generated profile (+ check in .js file)

joshspicer commented on build/lib/policies.ts:
```diff
@@ -590,6 +697,202 @@ function renderADML(appName: string, versions: string[], categories: Category[],
+function renderMacOSPolicy(policies: Policy[], translations: Translations) {
+	const appName = product.nameLong;
+	const bundleIdentifier = product.darwinBundleIdentifier;
+	const payloadUUID = '9F957DE8-6596-4CD6-A54E-D3A20F2B1C37';
```
It doesn't matter what these GUIDs are, they just need to be the same across releases (so we should not change them once we ship)

joshspicer replied:
Sounds good.  FWIW it's the bundleIdentifier + UUID that makes a profile unique, so profiles with this as-is would still be unique per quality

joshspicer replied:
https://github.com/microsoft/vscode/pull/241773/commits/a81273957e4e8b5b96bcb8d4a0594c162487f2d9. + distro PR

joshspicer committed:
tidy up output plists

joshspicer committed:
copy policy definitions

joshspicer committed:
implement the remaining renderProfileValue()

joshspicer committed:
remove unnecessary platform option

joshspicer commented on build/azure-pipelines/darwin/product-build-darwin.yml:
```diff
@@ -188,6 +193,21 @@ steps:
+          fi
+          mkdir -p "$APP_PATH/Contents/Resources/app/policies"
+          cp -r "$POLICIES_DIR/" "$APP_PATH/Contents/Resources/app/policies/"
+        displayName: Copy policy definitions
```
This bundles into the .app's resources, mirroring [how it works for Windows]().  Open to feedback on this approach

<img width="1037" alt="image" src="https://github.com/user-attachments/assets/699e10a6-7c57-437c-82c2-299f4c638a32" />

joshspicer replied:
chatted with Joao and moved copying to gulp script: https://github.com/microsoft/vscode/pull/241773/commits/a0d83794b90803fa4f0eda23273ef734b5dfd4fe

joshspicer committed:
copy to .app in vscode-darwin-$PLATFORM-min-ci gulp task

joaomoreno commented on build/gulpfile.vscode.js:
```diff
@@ -424,6 +424,10 @@ function packageTask(platform, arch, sourceFolderName, destinationFolderName, op
 				.pipe(replace('@@PRODNAME@@', product.nameLong))
 				.pipe(replace('@@APPNAME@@', product.applicationName))
 				.pipe(rename('bin/' + product.applicationName)));
+		} else if (platform === 'darwin') {
+			const policyDest = path.join(`${product.nameLong}.app`, 'Contents', 'Resources', 'app', 'policies');
+			result = es.merge(result, gulp.src('.build/policies/darwin/**', { base: '.build/policies/darwin' })
+				.pipe(rename(f => f.dirname = path.join(policyDest, f.dirname))));
```
Do it above the call to `electron()` so as to avoid hardcoding paths within the app bundle:

https://github.com/microsoft/vscode/blob/747b90cfcd486731a1295563423b1ee66dd9b613/build/gulpfile.vscode.js#L371-L378

joshspicer replied:
Ah yes thanks for the hint. That's much better.  Fixed in https://github.com/microsoft/vscode/pull/241773/commits/7c90a9a6c706ae54fb338614755d84095ebe44e1

<img width="827" alt="Screenshot 2025-03-13 at 11 21 31 AM" src="https://github.com/user-attachments/assets/d1d5e440-4931-4133-9ad6-0d84624ab714" />

joshspicer committed:
add darwinProfileUUID and darwinProfilePayloadUUID to product.json

joshspicer committed:
better way to package in gulp script

joshspicer committed:
bump distro commit (https://github.com/microsoft/vscode-distro/commit/c3ec5ba4852b5682b94358c92bf31484d2739db9)

joshspicer commented:
validated that packaging is working ok

- https://vscode.download.prss.microsoft.com/dbazure/download/insider/da9326ce3a543ad8ae4999640cab3dfa43745c4c/VSCode-darwin-arm64.zip
- https://vscode.download.prss.microsoft.com/dbazure/download/insider/da9326ce3a543ad8ae4999640cab3dfa43745c4c/VSCode-win32-arm64-1.99.0-insider.zip

<img width="742" alt="Screenshot 2025-03-13 at 1 44 01 PM" src="https://github.com/user-attachments/assets/459a3cbf-d958-41d1-a1ba-8ac7d609e9d7" />

<img width="730" alt="Screenshot 2025-03-13 at 1 44 15 PM" src="https://github.com/user-attachments/assets/2be0bb4b-c2d0-4463-9314-3435b579b756" />

joshspicer commented:
Ah, the universal macOS build has an [additional hashing](https://dev.azure.com/monacotools/Monaco/_build/results?buildId=324714&view=logs&j=fdf13662-433c-5ba9-9c13-a848a9d221fa&t=493be2f5-0db3-5cda-2f93-d236187783dc&l=10737) step that compares non-binary files in the result

```
2025-03-13T20:09:37.255Z electron-universal hashing /Users/runner/work/1/VSCode-darwin-arm64/Visual Studio Code - Insiders.app/Contents/Resources/app/policies/com.microsoft.VSCodeInsiders.mobileconfig
2025-03-13T20:09:37.255Z electron-universal hashing /Users/runner/work/1/VSCode-darwin-x64/Visual Studio Code - Insiders.app/Contents/Resources/app/policies/cs-cz/com.microsoft.VSCodeInsiders.plist
2025-03-13T20:09:37.256Z electron-universal hashing /Users/runner/work/1/VSCode-darwin-arm64/Visual Studio Code - Insiders.app/Contents/Resources/app/policies/cs-cz/com.microsoft.VSCodeInsiders.plist
2025-03-13T20:09:37.257Z electron-universal SHA for file Contents/Resources/app/policies/cs-cz/com.microsoft.VSCodeInsiders.plist does not match across builds ba009cb636170e045d3219941b240e62716fe8958462d326ee3da5d12ca43dea!=f2f67d90a872dd7f9980f5103e7243cd4f1d14469c14a542221e40a73bc039c2
Error: Expected all non-binary files to have identical SHAs when creating a universal build but "Contents/Resources/app/policies/cs-cz/com.microsoft.VSCodeInsiders.plist" did not
    at makeUniversalApp (/Users/runner/work/1/s/build/node_modules/vscode-universal-bundler/dist/cjs/index.js:120:23)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async main (/Users/runner/work/1/s/build/darwin/create-universal-app.js:33:5)
```

joshspicer committed:
Add bundles policy paths to macOS universal build 'filesToSkip'

Needs this when creating universal macOS app:

Expected all non-binary files to have identical SHAs when creating a universal build
but "Contents/Resources/app/policies/cs-cz/com.microsoft.VSCodeInsiders.plist" did not

See:
https://github.com/deepak1556/universal/blob/298a872f5f7d67eabb01d25464eef68edab14cd2/src/index.ts#L163-L172
